storage/mediator.go

```diff
@@ -0,0 +1,230 @@
+)
+
+type mediatorMetrics struct {
+	bootstrapStatus tally.Gauge
+	cleanupStatus   tally.Gauge
+	flushStatus     tally.Gauge
+	repairStatus    tally.Gauge
+}
+
+func newMediatorMetrics(scope tally.Scope) mediatorMetrics {
+	return mediatorMetrics{
+		bootstrapStatus: scope.Gauge("bootstrapped"),
+		cleanupStatus:   scope.Gauge("cleanup"),
+		flushStatus:     scope.Gauge("flush"),
+		repairStatus:    scope.Gauge("repair"),
+	}
+}
+
+type mediator struct {
+	sync.RWMutex
+	databaseBootstrapManager
+	databaseFileSystemManager
+	databaseTickManager
+	databaseRepairer
+
+	opts     Options
+	nowFn    clock.NowFn
+	sleepFn  sleepFn
+	metrics  mediatorMetrics
+	state    mediatorState
+	closedCh chan struct{}
+}
+
+func newMediator(database database, opts Options) (databaseMediator, error) {
+	scope := opts.InstrumentOptions().MetricsScope()
+	fsm, err := newFileSystemManager(database, scope)
+	if err != nil {
+		return nil, err
+	}
+
+	bsm := newBootstrapManager(database)
+	tm := newTickManager(database)
+
+	var repairer databaseRepairer
+	if opts.RepairEnabled() {
+		var err error
+		repairer, err = newDatabaseRepairer(database)
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		repairer = newNoopDatabaseRepairer()
+	}
+
+	return &mediator{
+		databaseBootstrapManager:  bsm,
+		databaseFileSystemManager: fsm,
+		databaseTickManager:       tm,
+		databaseRepairer:          repairer,
+		opts:                      opts,
+		nowFn:                     opts.ClockOptions().NowFn(),
+		sleepFn:                   time.Sleep,
+		metrics:                   newMediatorMetrics(scope),
+		state:                     mediatorNotOpen,
+		closedCh:                  make(chan struct{}),
+	}, nil
+}
+
+func (m *mediator) Open() error {
+	m.Lock()
+	defer m.Unlock()
+	if m.state != mediatorNotOpen {
+		return errMediatorAlreadyOpen
+	}
+	m.state = mediatorOpen
+	go m.reportLoop()
+	go m.ongoingTick()
+	m.databaseRepairer.Start()
+	return nil
+}
+
+func (m *mediator) IsBootstrapped() bool {
+	return m.databaseBootstrapManager.IsBootstrapped()
+}
+
+func (m *mediator) Bootstrap() error {
+	// NB(xichen): disable filesystem manager before we bootstrap to minimize
+	// the impact of file operations on node performance
+	fileOpInProgess := m.databaseFileSystemManager.Disable()
+	defer m.databaseFileSystemManager.Enable()
+
+	// Wait for in-flight file operations to finish before we proceed
+	// with bootstrapping
+	for fileOpInProgess {
+		m.sleepFn(fileOpCheckInterval)
+		fileOpInProgess = m.databaseFileSystemManager.IsRunning()
+	}
+
+	// Perform file operations after bootstrapping
+	err := m.databaseBootstrapManager.Bootstrap()
+	m.tickWithFileOp(0, false, true)
+	return err
+}
+
+func (m *mediator) Repair() error {
+	return m.databaseRepairer.Repair()
+}
+
+func (m *mediator) Close() error {
+	m.Lock()
+	defer m.Unlock()
+	if m.state == mediatorNotOpen {
+		return errMediatorNotOpen
+	}
+	if m.state == mediatorClosed {
+		return errMediatorAlreadyClosed
```

Comment:
If you wanted to save some paths you need to cover for code coverage you could combine these errors if you wanted, e.g. errMediatorCannotCloseNotOpen